l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol

```diff
@@ -0,0 +1,419 @@
+
+    ///  @inheritdoc IAssetHandler
+    /// @notice Used when the chain receives a transfer from L1 Shared Bridge and correspondingly mints the asset.
+    /// @param _chainId The chainId that the message is from.
+    /// @param _assetId The assetId of the asset being bridged.
+    /// @param _data The abi.encoded transfer data.
+    function bridgeMint(
+        uint256 _chainId,
+        bytes32 _assetId,
+        bytes calldata _data
+    ) external payable override onlyAssetRouter whenNotPaused {
+        // Either it was locked before, therefore is not zero, or it is sent from remote chain and standard erc20 will be deployed
+        // Note: if the chainBalance has not been migrated and is 0, then the assetId check will fail as the token's origin chain is not chain that sent the message.
+        // Note: after interop is implemented this will not work.
+        address token = tokenAddress[_assetId];
+        address receiver;
+        uint256 amount;
+        if (chainBalance[_chainId][token] > 0) {
+            (receiver, amount) = _bridgeMintNativeToken(_chainId, _assetId, _data);
+        } else {
+            (receiver, amount) = _bridgeMintBridgedToken(_chainId, _assetId, _data);
+        }
+        // solhint-disable-next-line func-named-parameters
+        emit BridgeMint(_chainId, _assetId, receiver, amount);
+    }
+
+    function _bridgeMintBridgedToken(
+        uint256 _originChainId,
+        bytes32 _assetId,
+        bytes calldata _data
+    ) internal virtual returns (address receiver, uint256 amount) {
+        // Either it was locked before, therefore is not zero, or it is sent from remote chain and standard erc20 will be deployed
+        address token = tokenAddress[_assetId];
+        bytes memory erc20Data;
+        address originToken;
+        // slither-disable-next-line unused-return
+        (, receiver, originToken, amount, erc20Data) = DataEncoding.decodeBridgeMintData(_data);
+
+        if (token == address(0)) {
+            token = _ensureTokenDeployed(_originChainId, _assetId, originToken, erc20Data);
+        }
+
+        IBridgedStandardToken(token).bridgeMint(receiver, amount);
+        emit BridgeMint(_originChainId, _assetId, receiver, amount);
+    }
+
+    function _bridgeMintNativeToken(
+        uint256 _originChainId,
+        bytes32 _assetId,
+        bytes calldata _data
+    ) internal returns (address receiver, uint256 amount) {
+        // Either it was locked before, therefore is not zero, or it is sent from remote chain and standard erc20 will be deployed
+        address token = tokenAddress[_assetId];
+        (amount, receiver) = abi.decode(_data, (uint256, address));
+        // Check that the chain has sufficient balance
+        if (chainBalance[_originChainId][token] < amount) {
+            revert InsufficientChainBalance();
+        }
+        chainBalance[_originChainId][token] -= amount;
+
+        if (token == ETH_TOKEN_ADDRESS) {
```

Comment:
You once called it `BASE_TOKEN_ADDRESS` and I really recommended to rename mainly because I saw places which looked like `ETH_TOKEN_ASSET_ID = NTV.encode(BASE_TOKEN_ADDRESS)` which could raise a lot of questions of sort whether the ETH_TOKEN_ASSET_ID is always correct under _any_ base token, etc. But now I am starting to understand a little bit better why you did the rename previously. 

Do you think a more consistent way would be to maintain the `BASE_TOKEN_ASSET_ID` as the immutable variable in each of the token vaults (it is set during genesis). Whenever someone tries to e.g. withdraw or move the base token asset id, we can use the standard `call` with value. Because right now (what I said in one of the previous comments) it is easy to have a collision between custom (non-eth) base token encoded with `ETH_TOKEN_ADDRESS`, which is ultimately incorrect IMHO.

Reply:
BASE_TOKEN_ASSET_ID

Yes, I think that would be a good idea

Reply:
```
    function createNewChain(
        uint256 _chainId,
        address _chainTypeManager,
        bytes32 _baseTokenAssetId,
        // solhint-disable-next-line no-unused-vars
        uint256 _salt,
        address _admin,
        bytes calldata _initData,
        bytes[] calldata _factoryDeps
```

but we already pass it, no?

Reply:
I meant the deployment of the BH contract, not the chains.